File: src/main/java/com/atlassian/migration/datacenter/core/fs/S3FilesystemMigrationService.java
/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.atlassian.migration.datacenter.core.fs;

import com.atlassian.jira.config.util.JiraHome;
import com.atlassian.migration.datacenter.core.exceptions.FileSystemMigrationFailure;
import com.atlassian.migration.datacenter.core.exceptions.InvalidMigrationStageError;
import com.atlassian.migration.datacenter.core.fs.download.s3sync.S3SyncFileSystemDownloadManager;
import com.atlassian.migration.datacenter.core.fs.reporting.DefaultFileSystemMigrationReport;
import com.atlassian.migration.datacenter.core.util.MigrationRunner;
import com.atlassian.migration.datacenter.dto.Migration;
import com.atlassian.migration.datacenter.spi.MigrationService;
import com.atlassian.migration.datacenter.spi.MigrationStage;
import com.atlassian.migration.datacenter.spi.fs.FilesystemMigrationService;
import com.atlassian.migration.datacenter.spi.fs.reporting.FileSystemMigrationReport;
import com.atlassian.scheduler.SchedulerService;
import com.atlassian.scheduler.SchedulerServiceException;
import com.atlassian.scheduler.config.JobConfig;
import com.atlassian.scheduler.config.JobId;
import com.atlassian.scheduler.config.JobRunnerKey;
import com.atlassian.scheduler.config.RunMode;
import com.atlassian.util.concurrent.Supplier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.services.s3.S3AsyncClient;

import javax.annotation.PostConstruct;
import java.nio.file.Path;
import java.nio.file.Paths;

import static com.atlassian.migration.datacenter.spi.MigrationStage.FS_MIGRATION_COPY;
import static com.atlassian.migration.datacenter.spi.fs.reporting.FilesystemMigrationStatus.DONE;
import static com.atlassian.migration.datacenter.spi.fs.reporting.FilesystemMigrationStatus.FAILED;
import static com.atlassian.migration.datacenter.spi.fs.reporting.FilesystemMigrationStatus.RUNNING;

public class S3FilesystemMigrationService implements FilesystemMigrationService {
    private static final Logger logger = LoggerFactory.getLogger(S3FilesystemMigrationService.class);

    private static final String OVERRIDE_UPLOAD_DIRECTORY = System.getProperty("com.atlassian.migration.datacenter.fs.overrideJiraHome", "");
    private static final String BUCKET_NAME = System.getProperty("S3_TARGET_BUCKET_NAME", "trebuchet-testing");

    private S3AsyncClient s3AsyncClient;
    private final JiraHome jiraHome;
    private final MigrationService migrationService;
    //private final SchedulerService schedulerService;
    private final MigrationRunner migrationRunner;
    private final S3SyncFileSystemDownloadManager fileSystemDownloadManager;
    private Supplier<S3AsyncClient> s3AsyncClientSupplier;

    private FileSystemMigrationReport report;
    private FilesystemUploader fsUploader;

    public S3FilesystemMigrationService(Supplier<S3AsyncClient> s3AsyncClientSupplier,
                                        JiraHome jiraHome,
                                        S3SyncFileSystemDownloadManager fileSystemDownloadManager,
                                        MigrationService migrationService,
                                        MigrationRunner migrationRunner) {
        this.s3AsyncClientSupplier = s3AsyncClientSupplier;
        this.jiraHome = jiraHome;
        this.migrationService = migrationService;
        this.migrationRunner = migrationRunner;
        this.fileSystemDownloadManager = fileSystemDownloadManager;

        this.report = new DefaultFileSystemMigrationReport();
    }

    @PostConstruct
    public void postConstruct() {
        this.s3AsyncClient = this.s3AsyncClientSupplier.get();
    }

    @Override
    public boolean isRunning() {
        return this.migrationService.getCurrentStage().equals(MigrationStage.FS_MIGRATION_COPY_WAIT);
    }

    @Override
    public FileSystemMigrationReport getReport() {
        return report;
    }

    @Override
    public Boolean scheduleMigration() throws InvalidMigrationStageError {
        Migration currentMigration = this.migrationService.getCurrentMigration();
        if (currentMigration.getStage() != FS_MIGRATION_COPY) {
            throw new InvalidMigrationStageError(
                    String.format(
                            "Cannot start filesystem migration as the system is not ready. Required state should be %s but is %s",
                            FS_MIGRATION_COPY,
                            currentMigration.getStage()));
        }

        JobId jobId = getScheduledJobId();
        S3UploadJobRunner jobRunner = new S3UploadJobRunner(this);

        boolean result = migrationRunner.runMigration(jobId, jobRunner);

        if (!result) {
            migrationService.error();
        }
        return result;
    }

    /**
     * Start filesystem migration to S3 bucket. This is a blocking operation and should be started from ExecutorService
     * or preferably from ScheduledJob
     */
    @Override
    public void startMigration() throws InvalidMigrationStageError {
        logger.trace("Beginning migration. Uploading shared home dir {} to S3 bucket {}", getSharedHomeDir(), getS3Bucket());
        if (isRunning()) {
            logger.warn("Filesystem migration is currently in progress, aborting new execution.");
            return;
        }

        s3AsyncClient = this.s3AsyncClientSupplier.get();
        report = new DefaultFileSystemMigrationReport();

        migrationService.transition(MigrationStage.FS_MIGRATION_COPY_WAIT);
        report.setStatus(RUNNING);

        Crawler homeCrawler = new DirectoryStreamCrawler(report);

        S3UploadConfig s3UploadConfig = new S3UploadConfig(getS3Bucket(), s3AsyncClient, getSharedHomeDir());
        Uploader s3Uploader = new S3Uploader(s3UploadConfig, report);

        fsUploader = new FilesystemUploader(homeCrawler, s3Uploader);

        logger.info("commencing upload of shared home");
        try {
            fsUploader.uploadDirectory(getSharedHomeDir());

            logger.info("upload of shared home complete. commencing shared home download");
            fileSystemDownloadManager.downloadFileSystem();

            report.setStatus(DONE);

            logger.info("Completed file system migration. Transitioning to next stage.");
            migrationService.transition(MigrationStage.OFFLINE_WARNING);
        } catch (FileSystemMigrationFailure e) {
            logger.error("Encountered critical error during file system migration");
            report.setStatus(FAILED);
            migrationService.error();
        }
    }

    @Override
    public void abortMigration() throws InvalidMigrationStageError {
        // We always try to remove scheduled job if the system is in inconsistent state
        migrationRunner.abortJobIfPresesnt(getScheduledJobId());

        if (!isRunning() || fsUploader == null) {
            throw new InvalidMigrationStageError(String.format("Invalid migration stage when cancelling filesystem migration: %s", migrationService.getCurrentStage()));
        }

        logger.warn("Aborting running filesystem migration");
        fsUploader.abort();
        report.setStatus(FAILED);

        migrationService.error();
    }

    private String getS3Bucket() {
        return BUCKET_NAME;
    }

    private JobId getScheduledJobId() {
        return JobId.of(S3UploadJobRunner.KEY + migrationService.getCurrentMigration().getID());
    }

    private Path getSharedHomeDir() {
        if (!OVERRIDE_UPLOAD_DIRECTORY.equals("")) {
            return Paths.get(OVERRIDE_UPLOAD_DIRECTORY);
        }
        return jiraHome.getHome().toPath();
    }
}

File: src/main/java/com/atlassian/migration/datacenter/core/fs/S3FilesystemMigrationService.kt
/*
 * Copyright 2020 Atlassian
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.atlassian.migration.datacenter.core.fs

import com.atlassian.jira.config.util.JiraHome
import com.atlassian.migration.datacenter.core.exceptions.FileUploadException
import com.atlassian.migration.datacenter.core.exceptions.InvalidMigrationStageError
import com.atlassian.migration.datacenter.core.fs.download.s3sync.S3SyncFileSystemDownloadManager
import com.atlassian.migration.datacenter.core.fs.download.s3sync.S3SyncFileSystemDownloader.CannotLaunchCommandException
import com.atlassian.migration.datacenter.core.fs.reporting.DefaultFileSystemMigrationReport
import com.atlassian.migration.datacenter.spi.MigrationService
import com.atlassian.migration.datacenter.spi.MigrationStage
import com.atlassian.migration.datacenter.spi.fs.FilesystemMigrationService
import com.atlassian.migration.datacenter.spi.fs.reporting.FileSystemMigrationReport
import com.atlassian.migration.datacenter.spi.fs.reporting.FilesystemMigrationStatus
import com.atlassian.scheduler.SchedulerService
import com.atlassian.scheduler.SchedulerServiceException
import com.atlassian.scheduler.config.JobConfig
import com.atlassian.scheduler.config.JobId
import com.atlassian.scheduler.config.JobRunnerKey
import com.atlassian.scheduler.config.RunMode
import com.atlassian.util.concurrent.Supplier
import org.slf4j.LoggerFactory
import software.amazon.awssdk.services.s3.S3AsyncClient
import java.nio.file.Path
import java.nio.file.Paths
import javax.annotation.PostConstruct

class S3FilesystemMigrationService(
    private val s3AsyncClientSupplier: Supplier<S3AsyncClient>,
    private val jiraHome: JiraHome,
    private val fileSystemDownloadManager: S3SyncFileSystemDownloadManager,
    private val migrationService: MigrationService,
    private val schedulerService: SchedulerService
) : FilesystemMigrationService {
    private var s3AsyncClient: S3AsyncClient? = null
    private var report: FileSystemMigrationReport = DefaultFileSystemMigrationReport()
    private var fsUploader: FilesystemUploader? = null

    @PostConstruct
    fun postConstruct() {
        s3AsyncClient = s3AsyncClientSupplier.get()
    }

    override fun isRunning(): Boolean {
        return migrationService.currentStage == MigrationStage.WAIT_FS_MIGRATION_COPY
    }

    override fun getReport(): FileSystemMigrationReport {
        return report
    }

    @Throws(InvalidMigrationStageError::class)
    override fun scheduleMigration(): Boolean {
        val currentMigration = migrationService.currentMigration
        if (currentMigration.stage != MigrationStage.FS_MIGRATION_COPY) {
            throw InvalidMigrationStageError(
                "Cannot start filesystem migration as the system is not ready. " +
                    "Required state should be ${MigrationStage.FS_MIGRATION_COPY} but is ${currentMigration.stage}"
            )
        }
        val runnerKey = JobRunnerKey.of(S3UploadJobRunner.Companion.KEY)
        val jobId = getScheduledJobId()
        logger.info("Starting filesystem migration")
        if (schedulerService.getJobDetails(jobId) != null) {
            logger.warn("Tried to schedule file system migration while job already exists")
            return false
        }
        // TODO: Can the job runner be injected? It has no state
        schedulerService.registerJobRunner(runnerKey, S3UploadJobRunner(this))
        logger.info("Registered new job runner for S3")
        val jobConfig = JobConfig.forJobRunnerKey(runnerKey)
            .withSchedule(null) // run now
            .withRunMode(RunMode.RUN_ONCE_PER_CLUSTER)
        try {
            logger.info("Scheduling new job for S3 upload runner")
            schedulerService.scheduleJob(jobId, jobConfig)
        } catch (e: SchedulerServiceException) {
            logger.error("Exception when scheduling S3 upload job", e)
            schedulerService.unscheduleJob(jobId)
            migrationService.error()
            return false
        }
        return true
    }

    /**
     * Start filesystem migration to S3 bucket. This is a blocking operation and should be started from ExecutorService
     * or preferably from ScheduledJob
     */
    @Throws(InvalidMigrationStageError::class)
    override fun startMigration() {
        logger.trace(
            "Beginning migration. Uploading shared home dir {} to S3 bucket {}",
            getSharedHomeDir(),
            getS3Bucket()
        )
        if (isRunning()) {
            logger.warn("Filesystem migration is currently in progress, aborting new execution.")
            return
        }
        s3AsyncClient = s3AsyncClientSupplier.get()
        report = DefaultFileSystemMigrationReport()
        migrationService.transition(MigrationStage.FS_MIGRATION_COPY, MigrationStage.WAIT_FS_MIGRATION_COPY)
        report.status = FilesystemMigrationStatus.RUNNING
        val homeCrawler: Crawler = DirectoryStreamCrawler(report)
        val s3UploadConfig = S3UploadConfig(getS3Bucket(), s3AsyncClientSupplier.get(), getSharedHomeDir())
        val s3Uploader: Uploader = S3Uploader(s3UploadConfig, report)
        fsUploader = FilesystemUploader(homeCrawler, s3Uploader)
        logger.trace("commencing upload of shared home")
        try {
            fsUploader!!.uploadDirectory(getSharedHomeDir())
        } catch (e: FileUploadException) {
            logger.error("Caught exception during upload; check report for details.", e)
        }
        if (report.status != FilesystemMigrationStatus.FAILED) {
            logger.trace("upload of shared home complete. commencing shared home download")
            try {
                fileSystemDownloadManager.downloadFileSystem()
                report.status = FilesystemMigrationStatus.DONE
            } catch (e: CannotLaunchCommandException) {
                report.status = FilesystemMigrationStatus.FAILED
                logger.error("unable to launch s3 sync ssm command", e)
            }
        }
        if (report.status == FilesystemMigrationStatus.DONE) {
            logger.trace("Completed file system migration. Transitioning to next stage.")
            migrationService.transition(MigrationStage.WAIT_FS_MIGRATION_COPY, MigrationStage.OFFLINE_WARNING)
        } else if (report.status == FilesystemMigrationStatus.FAILED) {
            logger.error("Encountered error during file system migration. Transitioning to error state.")
            migrationService.error()
            report.status = FilesystemMigrationStatus.DONE
        }
    }

    @Throws(InvalidMigrationStageError::class)
    override fun abortMigration() { // we always try to remove scheduled job if the system is in inconsistent state
        schedulerService.getJobDetails(getScheduledJobId())?.let {
            schedulerService.unscheduleJob(getScheduledJobId())
            logger.info("Removed scheduled filesystem migration job")
        }
        if (!isRunning() || fsUploader == null) {
            throw InvalidMigrationStageError(
                String.format(
                    "Invalid migration stage when cancelling filesystem migration: %s",
                    migrationService.currentStage
                )
            )
        }
        logger.warn("Aborting running filesystem migration")
        fsUploader!!.abort()
        report.status = FilesystemMigrationStatus.FAILED
        migrationService.error()
    }

    private fun getS3Bucket(): String {
        return BUCKET_NAME
    }

    private fun getScheduledJobId(): JobId {
        return JobId.of(S3UploadJobRunner.KEY + migrationService.currentMigration.id)
    }

    private fun getSharedHomeDir(): Path {
        return if (OVERRIDE_UPLOAD_DIRECTORY != "") {
            Paths.get(OVERRIDE_UPLOAD_DIRECTORY)
        } else jiraHome.home.toPath()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(S3FilesystemMigrationService::class.java)
        private val OVERRIDE_UPLOAD_DIRECTORY =
            System.getProperty("com.atlassian.migration.datacenter.fs.overrideJiraHome", "")
        private val BUCKET_NAME = System.getProperty("S3_TARGET_BUCKET_NAME", "trebuchet-testing")
    }
}